primo/utils/override_utils.py

```diff
@@ -0,0 +1,293 @@
+        for cluster_id, groups in self._plugged_list.groupby("Project"):
+            well_id = list(groups["API Well Number"])
+            num_well = len(well_id)
+            groups["coords"] = list(zip(groups.Latitude, groups.Longitude))
+            distance_metric_distance = haversine_vector(
+                groups["coords"].to_list(),
+                groups["coords"].to_list(),
+                unit=Unit.MILES,
+                comb=True,
+            )
+            well_distance = {}
+            for well_1 in range(num_well - 1):
+                for well_2 in range(well_1 + 1, num_well):
+                    well_distance = distance_metric_distance[well_1][well_2]
+                    if well_distance > distance_threshold:
+                        distance_violations[
+                            (cluster_id, well_id[well_1], well_id[well_2])
+                        ] = [
+                            cluster_id,
+                            well_id[well_1],
+                            well_id[well_2],
+                            well_distance,
+                        ]
+
+        return distance_violations
+
+    def assess_feasibility(self) -> bool:
+        """
+        Assesses whether current set of selections is feasible
+        """
+        if self.assess_budget() > 0:
+            return False
+
+        if self.assess_dac() > 0:
+            return False
+
+        if self.assess_owner_well_count():
+            return False
+
+        if self.assess_distances():
+            return False
+
+        return True
+
+    def backfill(self) -> List[str]:
+        """
+        If current selections (either due to manual overrides or termination of solver
+        prior to optimality)
+        leave room for budget, the method returns a list of candidates that can
+        be added to the plugging projects without violating any constraints
+
+        Parameters:
+        ----------
+        None
+
+        Returns:
+        --------
+        List of Well API numbers that can be added in the project
+        """
+        additions = []
+        if self.assess_feasibility() is False:
+            # The
+            return additions
+
+        # Sort by descending order
+        candidates = self.well_df.sort_values("Priority Score [0-100]", ascending=False)
+
+        # Only include those candidates not considered for plugging
+        candidates = candidates[
+            ~candidates["API Well Number"].isin(self._plugged_list["API Well Number"])
+        ]
+
+        # Remove candidates in removal list
+        candidates = candidates[~candidates["API Well Number"].isin(self.wells_removed)]
+        # Stick to those projects that already exist
+        existing_projects = set(self._plugged_list["Project"])
+        candidates = candidates[candidates["Project"].isin(existing_projects)]
+
+        for well_id in candidates["API Well Number"]:
+            self._add_well(well_id)
```

Comment:
We might need to clarify the function of the backfill feature. Do we also provide backfill candidates to users when some of the wells added to the suggested projects violate the constraint?

Reply:
At the moment, we don't! This is clarified in the check in line 268 (and the added comment).